src/net/stackerdb/bits.rs

```diff
@@ -0,0 +1,123 @@
+
+use stacks_common::types::chainstate::{ConsensusHash, StacksAddress, StacksPrivateKey};
+
+use stacks_common::types::PrivateKey;
+
+use crate::net::stackerdb::ChunkMetadata;
+
+use crate::chainstate::stacks::StacksPublicKey;
+
+use stacks_common::util::secp256k1::MessageSignature;
+
+impl ChunkMetadata {
+    /// Get the digest to sign that authenticates this chunk data and metadata
+    fn sighash(&self) -> Sha512Trunc256Sum {
+        let mut bytes = vec![];
+        bytes.extend_from_slice(&self.rc_consensus_hash.0);
+        bytes.extend_from_slice(&self.chunk_id.to_be_bytes());
+        bytes.extend_from_slice(&self.chunk_version.to_be_bytes());
+        bytes.extend_from_slice(&self.data_hash.0);
+
+        Sha512Trunc256Sum::from_data(&bytes)
+    }
+
+    /// Sign this chunk metadata, committing to rc_consensus_hash, chunk_id, chunk_version, and
+    /// data_hash.
+    pub fn sign(&mut self, privkey: &StacksPrivateKey) -> Result<(), net_error> {
+        let sigh = self.sighash();
+        let sig = privkey
+            .sign(&sigh.0)
+            .map_err(|se| net_error::SigningError(se.to_string()))?;
+
+        self.signature = sig;
+        Ok(())
+    }
+
+    /// Verify that a given principal signed this chunk metadata.
+    /// Note that the address version is ignored.
+    pub fn verify(&self, principal: &StacksAddress) -> Result<bool, net_error> {
+        let sigh = self.sighash();
+        let pubk = StacksPublicKey::recover_to_pubkey(sigh.as_bytes(), &self.signature)
+            .map_err(|ve| net_error::VerifyingError(ve.to_string()))?;
+
+        let pubkh = Hash160::from_node_public_key(&pubk);
+        Ok(pubkh == principal.bytes)
+    }
+}
+
+/// Helper methods for StackerDBChunkData messages
+impl StackerDBChunkData {
+    /// Create a new StackerDBChunkData instance.
+    pub fn new(chunk_id: u32, chunk_version: u32, data: Vec<u8>) -> StackerDBChunkData {
+        StackerDBChunkData {
+            chunk_id,
+            chunk_version,
+            sig: MessageSignature::empty(),
+            data,
+        }
+    }
+
+    /// Calculate the hash of the chunk bytes.  This is the SHA512/256 hash of the data.
+    pub fn data_hash(&self) -> Sha512Trunc256Sum {
+        Sha512Trunc256Sum::from_data(&self.data)
+    }
+
+    /// Create an owned ChunkMetadata describing the metadata of this chunk.
+    pub fn get_chunk_metadata(&self, rc_consensus_hash: ConsensusHash) -> ChunkMetadata {
+        ChunkMetadata {
+            rc_consensus_hash,
+            chunk_id: self.chunk_id,
+            chunk_version: self.chunk_version,
+            data_hash: self.data_hash(),
+            signature: self.sig.clone(),
+        }
+    }
+
+    /// Sign this given chunk data message with the given private key.  Binds it to a particular
+    /// reward cycle in a particular PoX history (identified by `rc_consensus_hash`).
+    /// Fails if the underlying signing library fails.
```

Comment:
Can you add a note to the rustdoc that the signature will be stored in `self.signature`?

Reply:
Yup